### PR TITLE
Fixing reshape op so it supports reshaping of scalars

### PR DESCRIPTION
**Pull request by ajakovljevicTT**
As described in https://github.com/tenstorrent/tt-mlir/issues/1306, tt-xla produces some reshapes on scalars which are currently not supported by our stableHLO to TTIR dialect lowering (issue https://github.com/tenstorrent/tt-mlir/issues/1317). In addition, tt-metal currently throws an error when trying to reshape to a 1-dimensional shape (issue https://github.com/tenstorrent/tt-metal/pull/15201). 

To fix both of these issues and fully support scalar testing on tt-xla, I've ingested some of the changes from @mmanzoorTT 's change (https://github.com/tenstorrent/tt-mlir/pull/1319) - will merge this PR after it. In addition, I've added that the reshape lowering pass omits the `ttir.reshape` op completely if the types of the input op and the target shape are the same, which will fix the unnecessary calls to the ttnn reshape op from tt-xla.

Fixes https://github.com/tenstorrent/tt-mlir/issues/1306

Edit: In talks with @sdjordjevicTT, he has pointed out the existence of the folder for transform passes for TTIR ops, so I have changed my pass to live there.

### CODE REVIEW

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 1]
```diff
@@ -202,13 +203,16 @@ class StableHLOToTTIRReshapeOpConversionPattern
...
-    rewriter.replaceOpWithNewOp<mlir::tt::ttir::ReshapeOp>(
+    auto new_reshape_op = rewriter.replaceOpWithNewOp<mlir::tt::ttir::ReshapeOp>(
         srcOp, getTypeConverter()->convertType(outputTensor.getType()),
-        srcOp->getOperand(0), outputTensor, new_shape_attr,
+        adaptor.getOperand(), outputTensor, new_shape_attr,
         rewriter.getArrayAttr(
             SmallVector<Attribute>(adaptor.getOperands().size() + 1,
                                    rewriter.getAttr<OperandConstraintAttr>(
                                        OperandConstraint::AnyDeviceTile))));
+    if (new_reshape_op.getType() == new_reshape_op->getOperand(0).getType()) {
```
Comment (kmitrovicTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:213]: A comment would be nice to explain this non-obvious statement.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 2]
```diff
@@ -19,6 +19,7 @@
 #include "ttmlir/Dialect/TTIR/IR/TTIROps.h"
 
 #include <llvm/ADT/APFloat.h>
+#include <mlir/Dialect/Func/IR/FuncOps.h>
```
Comment (kmitrovicTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:22]: Needed?

**Review by kmitrovicTT — APPROVED** [review]

**Comment by kmitrovicTT** [comment]
Run

```
> pre-commit install
> pre-commit run --all-files
```

to fix failing check.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 2 above)
Comment (ajakovljevicTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: You are correct, not needed. Fixed in the next commit.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 1 above)
Comment (ajakovljevicTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:216]: Fixed in the next commit.

**Comment by mmanzoorTT** [comment]
There is another PR (PR #1252 by @uazizTT) for handling the same issue for `broadcast` op (remove redundant ops). These two PRs are handling the same problem in two different location. Redundant `broadcast` op is removed in TTIR->TTIR conversion pass, whereas redundant `reshape` op is removed in StableHLO->TTIR conversion pass. These two PRs should remove redundant ops in a consistent way.

**Comment by mmanzoorTT** [comment]
Please add some tests for these changes.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecompositionPass.cpp [hunk 3]
```diff
@@ -50,6 +50,10 @@ struct TTIRToTTIRDecompositionPass
     target.addIllegalOp<ttir::ConvolutionOp>();
     target.addIllegalOp<ttir::GetDimensionSizeOp>();
     target.addIllegalOp<ttir::PoolingOp>();
+    target.addDynamicallyLegalOp<ttir::ReshapeOp>([](ttir::ReshapeOp op) {
```
Comment (uazizTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecompositionPass.cpp:53]: Is there a need for this? Do you not want to design it in such a way that this reshape op is either deleted specifically or removed by remove-dead-values pass?

Comment (ajakovljevicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecompositionPass.cpp:53]: As I have implemented the pattern inside the `TTIRtoTTIRDecompositionPass`, I have to mark the ReshapeOp conditionally illegal, so that the pattern will actually try to match the op. The other possible solution in to make a transform pass from scratch, as was done in https://github.com/tenstorrent/tt-mlir/pull/1252. Since we currently do not have a dedicated transformations pass, I put it here, but I am open to changing that and making a transformations pass folder.

Comment (ajakovljevicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecompositionPass.cpp]: Update: In talks with @sdjordjevicTT, he has pointed out the existence of the folder for transform passes for TTIR ops, so I have changed my pass to live there.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecompositionPass.cpp [hunk 4]
```diff
@@ -50,7 +50,6 @@ struct TTIRToTTIRDecompositionPass
     target.addIllegalOp<ttir::ConvolutionOp>();
     target.addIllegalOp<ttir::GetDimensionSizeOp>();
     target.addIllegalOp<ttir::PoolingOp>();
-    target.addIllegalOp<ttir::GatherOp>();
```
Comment (mmanzoorTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecompositionPass.cpp:53]: This seems irrelevant to your PR.

Comment (ajakovljevicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecompositionPass.cpp]: My mistake, fixed now.

File: test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir [hunk 5]
```diff
@@ -0,0 +1,10 @@
+// RUN: ttmlir-opt --ttir-reshape-fold %s| FileCheck %s
+#any_device_tile = #tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>
+module @jit_ravel attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
```
Comment (kmitrovicTT) [inline test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir:3]: Do you need these attributes? Is jit_ravel a good name?

File: lib/Dialect/TTIR/Transforms/Reshape.cpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                                PatternRewriter &rewriter) const final {
+    if (op.getType() == op->getOperand(0).getType()) {
+      rewriter.replaceOp(op, op->getOperand(0));
+      return success();
+    }
+    return failure();
+  }
+};
+
+class TTIRReshapeFold : public impl::TTIRReshapeFoldBase<TTIRReshapeFold> {
```
Comment (sdjordjevicTT) [inline lib/Dialect/TTIR/Transforms/Reshape.cpp:33]: @azecevicTT This seems like a good candidate for a canonicalization...

Comment (azecevicTT) [reply lib/Dialect/TTIR/Transforms/Reshape.cpp:33]: Yeah, these kind of things should be covered with canonicalization in the future. TableGen seems like a more suitable option for these kind of simple patterns, because it supports a proper pattern matching.

File: lib/Dialect/TTIR/Transforms/Reshape.cpp [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include <mlir/Transforms/GreedyPatternRewriteDriver.h>
+
+namespace mlir::tt::ttir {
+#define GEN_PASS_DEF_TTIRRESHAPEFOLD
+#include "ttmlir/Dialect/TTIR/Transforms/Passes.h.inc"
+
+//===----------------------------------------------------------------------===//
+// Constant as fill pass
+//===----------------------------------------------------------------------===//
```
Comment (uazizTT) [inline lib/Dialect/TTIR/Transforms/Reshape.cpp:17]: Please remove this header

Comment (ajakovljevicTT) [reply lib/Dialect/TTIR/Transforms/Reshape.cpp:17]: Fixed.

File: test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir (hunk 5 above)
Comment (ajakovljevicTT) [reply test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir]: You are correct, fixed now.

File: test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir [hunk 8]
```diff
@@ -0,0 +1,10 @@
+// RUN: ttmlir-opt --ttir-reshape-fold %s| FileCheck %s
+#any_device_tile = #tt.operand_constraint<dram|l1|tile|any_device_tile>
+module @reshape_test attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
```
Comment (svuckovicTT) [inline test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir:3]: Do you need these attributes?

File: lib/Dialect/TTIR/Transforms/Reshape.cpp (hunk 6 above)
Comment (LPanosTT) [reply lib/Dialect/TTIR/Transforms/Reshape.cpp:33]: @azecevicTT @sdjordjevicTT Something to take a look at for future patterns: https://mlir.llvm.org/docs/PDLL/. MLIR now supports defining patterns using PDLL in addition to tablegen. Somewhere in that page they give some reasoning as to why they made PDLL instead of just only having tablegen.

File: test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir (hunk 8 above)
Comment (ajakovljevicTT) [reply test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir:3]: copies from other tests, such as t`est/ttmlir/Dialect/TTIR/slice/slice_tests_positive.mlir` to reflect the ttir code generated from stable hlo. In addition, `ttir.reshape` requires an `operand_constraints` value.

Comment (ajakovljevicTT) [reply test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir]: Correct, addressed the comment.

File: test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir [hunk 9]
```diff
@@ -0,0 +1,10 @@
+// RUN: ttmlir-opt --ttir-reshape-fold %s| FileCheck %s
+#any_device_tile = #tt.operand_constraint<dram|l1|tile|any_device_tile>
+module @reshape_test {
```
Comment (svuckovicTT) [inline test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir:3]: Can you add a comment above the test, noting what the test is testing? We should try to be as targeted as possible, and I've started asking everyone to be explicit about what they are testing.

File: lib/Dialect/TTNN/Pipelines/TTNNPipelines.cpp [hunk 10]
```diff
@@ -107,8 +112,16 @@ void createTTNNPipelineDeallocPassFromString(OpPassManager &pm,
...
+void createTTIRReshapeFoldPassFromString(OpPassManager &pm,
+                                         std::string options) {
+  auto optionsStruct =
+      TTIRToTTNNBackendPipelineOptions::createFromString(options);
+  createTTNNPipelineDeallocPass(pm, *optionsStruct);
+}
+
 void createTTIRToTTNNBackendPipeline(
     OpPassManager &pm, const TTIRToTTNNBackendPipelineOptions &options) {
+  createTTIRReshapeFoldPass(pm, options);
```
Comment (svuckovicTT) [inline lib/Dialect/TTNN/Pipelines/TTNNPipelines.cpp:124]: Maybe this pass should live in `createTTNNPipelineTTIRPasses`?

@sdjordjevicTT can you comment on this? I'm not super familiar with the organization of these passes, but it seems to me that `createTTNNPipelineTTIRPasses` would be a better place for this pass?

File: test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir (hunk 9 above)
Comment (ajakovljevicTT) [reply test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir:4]: Addressed the comment.

File: lib/Dialect/TTNN/Pipelines/TTNNPipelines.cpp (hunk 10 above)
Comment (ajakovljevicTT) [reply lib/Dialect/TTNN/Pipelines/TTNNPipelines.cpp:124]: @sdjordjevicTT Did you have the time to take a quick look at this?

**Review by azecevicTT — COMMENTED** [review]
Curious why you decided to go with the whole pass, isn't the fold sufficient here?

**Comment by ajakovljevicTT** [comment]
> Curious why you decided to go with the whole pass, isn't the fold sufficient here?

@azecevicTT Just wanted to be in agreement with a PR that does a similar thing for broadcast https://github.com/tenstorrent/tt-mlir/pull/1252. Do you think it should be moved to a fold mechanism?

**Comment by azecevicTT** [comment]
> > Curious why you decided to go with the whole pass, isn't the fold sufficient here?
> 
> @azecevicTT Just wanted to be in agreement with a PR that does a similar thing for broadcast #1252. Do you think it should be moved to a fold mechanism?

I can see, but I think fold is ideal for simple rewritings like this (and broadcast too). In the future we will probably have 10s if not 100s of this kind of rewrting patterns, having a separate pass for all of them can become burdensome from a maintainability standpoint, so it will be best to use an existing infrastructure, or group them under one pass that will operate on all ops, at the moment fold seems like the best option.

**Comment by ajakovljevicTT** [comment]
> > > Curious why you decided to go with the whole pass, isn't the fold sufficient here?
> > 
> > 
> > @azecevicTT Just wanted to be in agreement with a PR that does a similar thing for broadcast #1252. Do you think it should be moved to a fold mechanism?
> 
> I can see, but I think fold is ideal for simple rewritings like this (and broadcast too). In the future we will probably have 10s if not 100s of this kind of rewrting patterns, having a separate pass for all of them can become burdensome from a maintainability standpoint, so it will be best to use an existing infrastructure, or group them under one pass that will operate on all ops, at the moment fold seems like the best option.

@azecevicTT Added golding and tested if it works on our current setup with the test in the PR. It seems to be working, so added the ommission as a fold instead of a pass. Thanks for pointing that out!

File: test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir [hunk 11]
```diff
@@ -0,0 +1,11 @@
+// RUN: ttmlir-opt --ttir-to-ttnn-backend-pipeline %s| FileCheck %s
```
Comment (azecevicTT) [inline test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir:1]: Since you are running it now with `--ttir-to-ttnn-backend-pipeline` just move it to the TTNN tests and rename it for the clarity, something like `Dialect/TTNN/reshape/reshape_folding_test.mlir`. Otherwise looks great!

**Review by azecevicTT — APPROVED** [review]

File: test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir (hunk 11 above)
Comment (ajakovljevicTT) [reply test/ttmlir/Dialect/TTIR/reshape/reshape_test.mlir]: Moved the test as you mentioned, thanks for looking into this!

**Review by sdjordjevicTT — APPROVED** [review]
Nice and elegant! Thanks.

File: test/ttmlir/Dialect/TTNN/reshape/reshape_folding_test.mlir [hunk 12]
```diff
@@ -0,0 +1,11 @@
+// RUN: ttmlir-opt --ttir-to-ttnn-backend-pipeline %s| FileCheck %s
+#any_device_tile = #tt.operand_constraint<dram|l1|tile|any_device_tile>
+// Tests if we fold when translating from "ttir.reshape" which is called on the two same shapes.
+module @reshape_test {
+  func.func @main(%arg0: tensor<1xi32>) -> (tensor<1xi32> {jax.result_info = ""}) {
+    %0 = tensor.empty() : tensor<1xi32>
+    %1 = "ttir.reshape"(%arg0, %0) <{operand_constraints = [#any_device_tile, #any_device_tile], shape = [1 : i32]}> : (tensor<1xi32>, tensor<1xi32>) -> tensor<1xi32>
+    // CHECK: return %arg0 : tensor<1xi32, #{{.*}}>
```
Comment (mmanzoorTT) [inline test/ttmlir/Dialect/TTNN/reshape/reshape_folding_test.mlir:8]: Please add a `CHECK-NOT` statement to verify that there is no `reshape` op in the generated graph.

**Review by mmanzoorTT — APPROVED** [review]

File: test/ttmlir/Dialect/TTNN/reshape/reshape_folding_test.mlir (hunk 12 above)
Comment (ajakovljevicTT) [reply test/ttmlir/Dialect/TTNN/reshape/reshape_folding_test.mlir:8]: Changed the test as you suggested.

File: test/ttmlir/Dialect/TTNN/reshape/reshape_folding_test.mlir [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+    // CHECK-NOT: %[[C:.*]] = "ttnn.reshape"[C:.*]]
```
Comment (mmanzoorTT) [inline test/ttmlir/Dialect/TTNN/reshape/reshape_folding_test.mlir:9]: This will check that `ttnn.reshape` is not present after previous `CHECK` statement (which is checking `return` statement). Please move it above line 8. So FileCheck verify it before the `return` statement and not afterwards.

**Review by vmilosevic — APPROVED** [review]

**Review by tapspatel — APPROVED** [review]
